### PR TITLE
Add a remixer_storage module that can convert RemixerItems into Json and back.

### PR DESCRIPTION
**Pull request by miguelandres**
-This new package converts variables and triggers to an intermediate object (StoredVariable) that maps directly to the Json scheme we want to use for storage and sync.
-There is a Gson Type Adapter for StoredVariable that allows it to convert to and from Json.
-SupportedDataType is an enum that implements the type-dependent conversion.

This Json package will be hooked up to the core through the implementation of Storage/sync interfaces, this is still not done (I don't wanna send a HUGE PR).

This is the first step for #9 and #10

### CODE REVIEW

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+package com.google.android.libraries.remixer.storage.json;
+
+import com.google.android.libraries.remixer.RemixerItem;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+class RemixerStatus {
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java]: documentation please even for private classes

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  Map<String, StoredVariable> map;
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java]: would like to see better name for this

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+      if (item instanceof Variable) {
+        Variable var = (Variable) item;
+        if (var.getVariableType() == Boolean.class) {
+          StoredVariable<Boolean> storage = new StoredVariable<>();
+          storage.dataType = BOOLEAN.getType();
+          storage.selectedValue = (Boolean) var.getSelectedValue();
+          return storage;
+        }
+      }
+      return null;
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: or should you throw?

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+  SupportedDataType(String type, ValueConverter valueConverter) {
+    this.type = type;
+    this.valueConverter = valueConverter;
+  }
+  private String type;
+
+  private ValueConverter valueConverter;
+
+  public String getType() {
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: the code is slightly difficult to follow because of the ambiguity of "type". It's used for both `String` and `Type`. Rather than String, can you just use `Type` here and in RemixerStatus/Storage? Will the equality checks still work out?

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+   * Creates a StoredVariable that represents the data in {@code item} if {@code item} is of this
+   * type. Returns null otherwise.
+   */
+  abstract StoredVariable<T> fromRemixerItem(RemixerItem item);
+
+  /**
+   * Deserializes a JsonElement that contains a StoredVariable.
+   */
+  StoredVariable<T> deserialize(JsonElement json) {
+    StoredVariable<T> result = createStoredVariable();
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java]: can you explain why you cannot just have this here:

```
StoredVariable<T> result = new StoredVariable<T>();
```

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java (hunk 1 above)
Comment (pingpongboss) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java]: Also I find RemixerStatus to be a confusing name

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+      }
+    }
+    if (storage == null) {
+      throw new UnsupportedOperationException(
+          "Cannot convert remixer item, maybe you forgot to add support for a new type?");
+    }
+    storage.key = item.getKey();
+    storage.title = item.getTitle();
+    return storage;
+
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java]: empty lines. format

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    @Override
+    public StoredVariable<Boolean> createStoredVariable() {
+      return new StoredVariable<>();
+    }
+  }),
+  COLOR("color", new ValueConverter<Integer>("color") {
+    @Override
+    public Integer parseValue(JsonElement element) {
+      JsonObject object = element.getAsJsonObject();
+      return Color.argb(
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java:68]: have you considered just sending and receiving the entire color as an `int` rather than 4 ints? What are the pros and cons? 

One pro is that you could consider COLOR and NUMBER as the same supported type. The fact that you use StoredVariable<Integer> for COLOR means that this should be possible.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 3 above)
Comment (pingpongboss) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: Since you throw in other parts of this PR, you should probably throw here

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+      return object;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    StoredVariable<Integer> fromRemixerItem(RemixerItem item) {
+      if (item instanceof Variable) {
+        Variable var = (Variable) item;
+        if (var.getVariableType() == Integer.class
+            && var.getLayoutId() == R.layout.color_list_variable_widget) {
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java:91]: you probably knew i was going to say this, but I don't love the implicit contract here and below:

```
var.getVariableType() == Integer.class
 +            && var.getLayoutId() == R.layout.color_list_variable_widget
```

Any way to make this explicit? May not belong in this PR.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+      if (item instanceof Trigger) {
+        StoredVariable<Void> storage = new StoredVariable<>();
+        storage.dataType = TRIGGER.getType();
+        return storage;
+      }
+      return null;
+    }
+  });
+
+  SupportedDataType(String type, ValueConverter valueConverter) {
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: move constructor below fields

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 4 above)
Comment (pingpongboss) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: hmm is it because of serialization/deserialization? If that's the case, I would reserve `type` for `Type`, and use something different like `dataType` for `String`

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+  }
+  private String type;
+
+  private ValueConverter valueConverter;
+
+  public String getType() {
+    return type;
+  }
+
+  public ValueConverter getValueConverter() {
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: for consistency, since this is a package-private class, declare fields are package-private and remove public getters

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 8 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java:91]: Since android doesn't have a color class this is as far as we can go with the current status

I don't love it either, I honestly think we may be getting rid of the layout id long term and just having a few really well supported layouts, and then making the type explicit via an enum.

I really hate the way variable type is being queried as well because it's really unsafe. It's quite unnerving.

This needs a larger refactor, which I'm thinking about in the backburner, but mostly being explicit about types, moving some of the SupportedDataType stuff back to core, for example.

Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java:91]: (Honestly a lot of things would be FAR better if Java didn't do type erasure :/)

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java (hunk 2 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java]: SerializedRemixerData?

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 3 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: Done

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 4 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: dataTypeSerializableString, added doc comments too.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java (hunk 5 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java]: You can, indeed. This code was refactored from outside where we didn't have the T param, now that doesn't make sense. thanks.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java (hunk 6 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java]: Done

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 7 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java:68]: I have argued in favor of using either that or the hex string but lost every time :) This needs to be compatible with the protocol in web/iOS/firebaseCloud as well.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java (hunk 1 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java]: Done

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 3 above)
Comment (miguelandres) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: Done

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java (hunk 5 above)
Comment (miguelandres) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java]: Done. This was refactored from before the deserialization had the <T> param, so it just stayed there accidentally, thanks

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 9 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: Done

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 10 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: Done.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java (hunk 2 above)
Comment (pingpongboss) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatus.java]: Yea much better 👍

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 7 above)
Comment (pingpongboss) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java:68]: Oh interesting, so you have cross-platform concerns to deal with. Very cool, thanks for explanation.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatusSerializer.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+class RemixerStatusSerializer
+    implements JsonSerializer<SerializableRemixerContents>, JsonDeserializer<SerializableRemixerContents> {
+
+  @Override
+  public SerializableRemixerContents deserialize(
+      JsonElement json, Type typeOfT, JsonDeserializationContext context)
+      throws JsonParseException {
+    Gson gson = GsonProvider.getInstance();
+    SerializableRemixerContents status = new SerializableRemixerContents();
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatusSerializer.java]: i would change the local var `status` name as well

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+ * full status of Remixer across the entire app.
+ */
+public class SerializableRemixerContents {
+
+  /**
+   * Mapping from Remixer Item ID to the item's representation in Serializable format.
+   *
+   * <p>Notice that while {@link com.google.android.libraries.remixer.Remixer#keyMap} contains more
+   * than one item per key, all of those items contain the same data (same value for variables), so
+   * we only keep one of them here.
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java:38]: each item contains a distinct callback right? How do you restore those?

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Set;
+
+/**
+ * A copy of the Remixer data structure in an easily serializable form.
+ *
+ * <p>This representation discards any runtime details (number of variables with the same key,
+ * callbacks, etc) to just focus on data. This object will be serialized to Json and back to get the
+ * full status of Remixer across the entire app.
+ */
+public class SerializableRemixerContents {
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java]: is this class public on purpose?

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+import com.google.android.libraries.remixer.RemixerItem;
+import java.util.List;
+
+/**
+ * This is an intermediate format that conforms to the agreed upon cross-platform storage and sync
+ * protocol.
+ *
+ * <p>This is never used for anything other than storage and syncing and is meant to be converted to
+ * a regular {@link com.google.android.libraries.remixer.RemixerItem} as soon as it's completely
+ * parsed.
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java:29]: I really like the javadoc here. Does this class have similar parallels with SerializableRemixerContents? May want to add some of this wording to that javadoc as well.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    result = 31 * result + (maxValue != null ? maxValue.hashCode() : 0);
+    result = 31 * result + (increment != null ? increment.hashCode() : 0);
+    return result;
+  }
+
+  /**
+   * Creates a Stored variable from a existing RemixerItem.
+   */
+  static StoredVariable fromRemixerItem(RemixerItem item) {
+    StoredVariable storage = null;
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java]: You define `StoredVariable variable` elsewhere which I think is a better name than `storage`

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+  private final String dataType;
+
+  ValueConverter(String dataType) {
+    this.dataType = dataType;
+  }
+
+  /**
+   * Returns an object of type T that holds the value in the current Json Element.
+   */
+  abstract T parseValue(JsonElement element);
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java:47]: should/could these 2 methods be marked @Nullable? Trigger returns null above and it would be helpful to know what happens in that case

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+
+    @Override
+    @SuppressWarnings("unchecked")
+    StoredVariable<Void> fromRemixerItem(RemixerItem item) {
+      if (item instanceof Trigger) {
+        StoredVariable<Void> storage = new StoredVariable<>();
+        storage.dataType = TRIGGER.getDataTypeSerializableString();
+        return storage;
+      }
+      return null;
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: throw here?

File: remixer_storage/src/test/java/com/google/android/libraries/remixer/storage/json/BooleanVariableSerializationTest.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(
+    sdk = 21,
+    manifest = "src/main/AndroidManifest.xml",
+    packageName = "com.google.android.libraries.remixer.ui")
```
Comment (pingpongboss) [inline remixer_storage/src/test/java/com/google/android/libraries/remixer/storage/json/BooleanVariableSerializationTest.java]: i'm curious how you determine the package name here. This test is in `remixer_storage`.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+    }
+  });
+
+  /**
+   * A string that represents this dataType when serialized. When deserializing it's used to
+   * identify the parsing logic to use.
+   */
```
Comment (pingpongboss) [inline remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: tiiiny nit: these 2 javadoc may be more useful on the accessor method rather than the field. Arguable.

File: remixer_storage/src/test/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContentsSerializationTest.java [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+public class SerializableRemixerContentsSerializationTest {
+
+  private Variable<Boolean> booleanVariable;
+  private ItemListVariable<Integer> colorListVariable;
+  private Variable<Integer> integerVariable;
+  private ItemListVariable<Integer> integerListVariable;
+  private RangeVariable rangeVariable;
+  private ItemListVariable<String> stringListVariable;
+  private Variable<String> stringVariable;
+  private SerializableRemixerContents status;
```
Comment (pingpongboss) [inline remixer_storage/src/test/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContentsSerializationTest.java]: you'll want to do a project-wide find&replace for "status"

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java (hunk 13 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java]: It is, I plan to use it from other packages, this will be the main entry point.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java (hunk 12 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java:38]: We do not restore them from here, either they get added at some point (and their value will copy the one stored here), or we just update the value if some change is detected after they are added (i.e. firebase change)

Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java:38]: If they are never added then the code that initializes the callback is never run, which is fine since this is basically a singleton and the variables are context scoped.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatusSerializer.java (hunk 11 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/RemixerStatusSerializer.java]: Done

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java (hunk 13 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContents.java]: I will make it package only until I need it.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java (hunk 14 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/StoredVariable.java:29]: Tried to explain these two issues there.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java (hunk 16 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/ValueConverter.java:47]: Marked nullable.

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 17 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: Done

File: remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java (hunk 19 above)
Comment (miguelandres) [reply remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/json/SupportedDataType.java]: Done

File: remixer_storage/src/test/java/com/google/android/libraries/remixer/storage/json/BooleanVariableSerializationTest.java (hunk 18 above)
Comment (miguelandres) [reply remixer_storage/src/test/java/com/google/android/libraries/remixer/storage/json/BooleanVariableSerializationTest.java]: packageName seems to be optional. Will remove it.

If internal google tools require it I will readd it.

Comment (miguelandres) [reply remixer_storage/src/test/java/com/google/android/libraries/remixer/storage/json/BooleanVariableSerializationTest.java]: (FWIW it should be com.google.android.libraries.remixer.storage if I were keeping it)

File: remixer_storage/src/test/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContentsSerializationTest.java (hunk 20 above)
Comment (miguelandres) [reply remixer_storage/src/test/java/com/google/android/libraries/remixer/storage/json/SerializableRemixerContentsSerializationTest.java]: Done
